File: drills/ts/harnesses/nc86_subsets_ii.ts
import * as solution from "./solution";

const show = (v: unknown) => JSON.stringify(v) ?? "undefined";

const sorted = (nums: number[]) =>
  solution
    .subsetsWithDup(nums)
    .map((subset) => [...subset].sort((a, b) => a - b).join(","))
    .sort();

export function run(): [string, string, string][] {
  if (typeof solution.subsetsWithDup !== "function") throw new Error("__signature_mismatch__");
  return [
    ["subsetsWithDup([1, 2, 2])", show(["", "1", "1,2", "1,2,2", "2", "2,2"]), show(sorted([1, 2, 2]))],
    ["subsetsWithDup([0])", show(["", "0"]), show(sorted([0]))],
    ["subsetsWithDup([])", show([""]), show(sorted([]))],
    ["subsetsWithDup([1, 1, 1])", show(["", "1", "1,1", "1,1,1"]), show(sorted([1, 1, 1]))],
    ["subsetsWithDup([4, 4, 4, 1, 4]) count", show(10), show(solution.subsetsWithDup([4, 4, 4, 1, 4]).length)],
  ];
}

File: drills/elixir/harnesses/nc86_subsets_ii.exs
sorted = fn nums ->
  nums
  |> Solution.subsets_with_dup()
  |> Enum.map(fn subset -> subset |> Enum.sort() |> Enum.join(",") end)
  |> Enum.sort()
end

[
  {"subsets_with_dup([1, 2, 2])", inspect(["", "1", "1,2", "1,2,2", "2", "2,2"]),
   inspect(sorted.([1, 2, 2]))},
  {"subsets_with_dup([0])", inspect(["", "0"]), inspect(sorted.([0]))},
  {"subsets_with_dup([])", inspect([""]), inspect(sorted.([]))},
  {"subsets_with_dup([1, 1, 1])", inspect(["", "1", "1,1", "1,1,1"]),
   inspect(sorted.([1, 1, 1]))},
  {"subsets_with_dup([4, 4, 4, 1, 4]) count", inspect(10),
   inspect(length(Solution.subsets_with_dup([4, 4, 4, 1, 4])))}
]
